src/base/misc/fatfs.c

```diff
@@ -417,27 +461,23 @@ int read_boot(fatfs_t *f, unsigned char *b)
 
   if(f->boot_sec) {
     memcpy(b, f->boot_sec, 0x200);
-    set_geometry(f, b);
-    if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {
-      bpb->v340.drive_number = f->drive_num;
-    } else {
-      b[0x1fd] = f->drive_num;
-    }
-    return 0;
-  }
+  } else {
+    // build v4 boot block for dosemu's boot code
 
-  build_boot_blk(f, b);
-  set_geometry(f, b);
+    build_boot_blk(f, b);
 
-  memcpy(b + 0x03, "DOSEMU10", 8);
+    memcpy(b + 0x03, "DOSEMU10", 8);
 
-  bpb->v400.drive_number = f->drive_num;
-  bpb->v400.flags = 0;
-  bpb->v400.signature = 0x29;
-  bpb->v400.serial_number = f->serial;
-  memcpy(bpb->v400.vol_label,  f->label, 11);
-  memcpy(bpb->v400.fat_type,
+    bpb->v400.drive_number = f->drive_num;
+    bpb->v400.flags = 0;
+    bpb->v400.signature = BPB_V400;
+    bpb->v400.serial_number = f->serial;
+    memcpy(bpb->v400.vol_label,  f->label, 11);
+    memcpy(bpb->v400.fat_type,
          f->fat_type == FAT_TYPE_FAT12 ? "FAT12   " : "FAT16   ", 8);
+  }
+
+  set_geometry(f, b);
```

Comment:
You assign drive number in set_geometry() now,
plus it is also assigned on line 471. I wonder why
moving the drive number to set_geometry()?